### PR TITLE
feat(ext/crypto): implement importKey (jwk) for RSA-PSS , RSA-OAEP and RSA-PKCS1-v1_5

### PR DESCRIPTION
**Pull request by SeanWykes**
towards #11690, implementing: 
- importKey(jwk) for RSA-PSS, RSASSA-PKCS1-v1_5, RSA-OAEP
- exportKey(jwk) for same algorithms.

### CODE REVIEW

**Comment by littledivy** [comment]
@SeanWykes pkcs#8 for all RSA algorithms is being done in https://github.com/denoland/deno/pull/11891

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1363]: This won't work. The key for a particular algorithm must be same throughout all the operations.

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1278]: Use the `spki` crate for this. https://docs.rs/spki/0.4.0/spki/

**Review by littledivy — CHANGES_REQUESTED** [review]

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:227]: The `rsa` crate already comes with a ASN.1 DER decoder/encoder. It's better to use those instead of rolling a JavaScript implementation.

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1210]: This should be a `DataError` according to step 7

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1226]: `jwk.n` can be undefined here. Move this after step 9.

**Review by littledivy — CHANGES_REQUESTED** [review]

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js:1363]: Sorry, I can't see the problem here. The code imports jwk, converting private to 'pkcs8' format and public to spki format for  use of the other op_crypto operations. Please elucidate.

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js:1278]: Ok. I think its best to wait on #11891, then include spki as an option in op_crypto_import_key.

File: ext/crypto/00_crypto.js
Comment (littledivy) [reply ext/crypto/00_crypto.js:1363]: Other RSA operations assume that the key is of type raw i.e `pkcs1` format. 
An `spki`/`pkcs8` key will fail eg: https://github.com/denoland/deno/blob/a95ca9dc70515240b1be7a12dbf686b1ebfb3490/ext/crypto/lib.rs#L275

Comment (SeanWykes) [reply ext/crypto/00_crypto.js:1150]: ok, PR#11880 altered this .. I had patched the operations to match both spki/pkcs8 .. will rework it.

Comment (littledivy) [reply ext/crypto/00_crypto.js]: Instead of having mixed type keys in storage. Can you just call `.to_pkcs1_der()` on the private key (once you have implemented the Rust op) and store it as pkcs1? It will make life much easier for the maintenance and eventually migrating to a new crypto storage.

Comment (SeanWykes) [reply ext/crypto/00_crypto.js]: reworked and merged. 

There will still be a problem with importing RSA Public keys via SPKI/JWK, since I am now using the KeyFormat enum to differentiate between private (Raw) and public (Spki). Ideally we would store public keys in Pkcs1 format, I think, so maybe need to include a new property in KeyData that indicates private/public. What do you think?

Comment (littledivy) [reply ext/crypto/00_crypto.js]: The `CryptoKey` object already has the `type` property which is associated to the key handle.

Comment (SeanWykes) [reply ext/crypto/00_crypto.js]: Sure, but that currently doesn't reach rust ops, for example, op_crypto_verify_key receives a VerifyArg with key: KeyData object but no CryptoKey. Verify must be able to distinguish between imported private and public keys.

Comment (littledivy) [reply ext/crypto/00_crypto.js]: Ah yes good point, we'll need to pass the `type` to these particular rust ops. It probably wasn't done before we weren't dealing with public keys directly.

Comment (SeanWykes) [reply ext/crypto/00_crypto.js]: Need to include a "keyType" property on KeyData, and copy from CryptoKey.type during import/generate .. ECDH and ECDSA will need this as well.

File: ext/crypto/lib.rs
Comment (littledivy) [inline ext/crypto/lib.rs:285]:
```suggestion
#[serde(rename_all = "camelCase")]
```

Minor nit. All `key_type` in JS side can now be renamed to `keyType`.

File: ext/crypto/lib.rs
Comment (littledivy) [inline ext/crypto/lib.rs:283]:
```suggestion
use crate::key::KeyType;
```
It's already defined in `ext/crypto/key.rs` ^_^

File: ext/crypto/lib.rs
Comment (littledivy) [inline ext/crypto/lib.rs:468]:
```suggestion
          RsaPublicKey::from_pkcs1_der(&*args.key.data)?
```

File: ext/crypto/lib.rs
Comment (littledivy) [inline ext/crypto/lib.rs:532]:
```suggestion
          RsaPublicKey::from_pkcs1_der(&*args.key.data)?
```

File: ext/crypto/lib.rs
Comment (littledivy) [inline ext/crypto/lib.rs:828]:
```suggestion
          RsaPublicKey::from_pkcs1_der(&*args.key.data)?
```

**Review by littledivy — CHANGES_REQUESTED** [review]

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1474]: `unwrapKey` and `wrapKey` are valid too

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:1156]:
```suggestion
                      "Hash algorithm not supported",
```

This can be extended by other specifications.

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js:1474]: Ok. I understood that the general idea with CryptoAPI was not to accept anything that was not yet implemented, such as wrapKey and unwrapKey. That's why I excluded them. I see however that pkcs8 import has them.

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js]: ok, expanded to all instances, PS/RS/RSA-OAEP

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js]: @littledivy - I'll create a new branch/PR for spki, converting in rust to a pkcs1 public key. Can probably include export as well .. and was wondering what are your thoughts on exportKey for spki (and jwk), since we'll need to parse the raw pkcs1 to extract the key components. Rust or js?

Going forward, I saw some comments in `op_crypto_import_key` about importing jwk .. is this on the to-do list?

File: ext/crypto/00_crypto.js
Comment (littledivy) [inline ext/crypto/00_crypto.js:221]: As pointed out in https://github.com/denoland/deno/pull/12022#discussion_r707008417 The [`rsa::pkcs1::ToRsaPublicKey`](https://docs.rs/pkcs1/0.2.4/pkcs1/trait.ToRsaPublicKey.html) and it's corresponding private key trait offers exactly what you need. Example for public key:
```rust
pub struct JWK { ... };

impl ToRsaPublicKey for JWK {
  fn to_pkcs1_der(&self) -> Result<pkcs1::RsaPublicKeyDocument> {
    let pub_key = pkcs1::RsaPublicKey { 
      modulus: decode(self.n),
      public_exponent: decode(self.e),
    };

    Ok(pub_key.to_der())
  }
}
```

File: ext/crypto/00_crypto.js
Comment (littledivy) [reply ext/crypto/00_crypto.js]: @SeanWykes Preferably using Rust since we already have the dependencies. See https://github.com/denoland/deno/pull/12022#discussion_r710707372 

> Going forward, I saw some comments in op_crypto_import_key about importing jwk .. is this on the to-do list?

It's on the todo but feel free to work on it!

Comment (littledivy) [reply ext/crypto/00_crypto.js]: I've opened a PR for exporting spki keys https://github.com/denoland/deno/pull/12114. Hopefully that can help you with importKey

**Comment by littledivy** [comment]
@SeanWykes Are you still working on this?

**Comment by SeanWykes** [comment]
@littledivy intend to get back to it this week, need some courage to deal with the conflicts from main branch ;)

**Comment by SeanWykes** [comment]
> @SeanWykes Are you still working on this?

@littledivy Please advise on what you consider the best way to pass the JSONWebKey hashmap into Rust. 

Options:
1. Declare a JWK pub struct with 8 x Option<ZeroCopyBuf> for RSA, 4 for EC + 1 for symmetric? This soon gets voluminous and messy. Base64->ArrayBuffer in js.
2. As (1) but pass B64 to Rust
3. Pass a Vec<(String,String)> to Rust and deal with convertion and validation there?
4. ???

What's the "Deno" way of doing things, that would align with future plans to move keystore into Rust?

Thanks

**Comment by bartlomieju** [comment]
@lucacasonato PTAL

**Comment by bartlomieju** [comment]
@SeanWykes could you please rebase the PR? I would love to land it for 1.17 in two weeks.

**Comment by SeanWykes** [comment]
> @SeanWykes could you please rebase the PR? I would love to land it for 1.17 in two weeks.
@bartlomieju - rebased and cleaned up.
@littledivy and @lucacasonato - ready for code review.

**Review by bartlomieju — COMMENTED** [review]
@SeanWykes thank you for the PR, it looks mostly good to me. I want to get review from @bnoordhuis and @lucacasonato before landing. 

As a heads up it looks to me that there are gonna be conflicts between this PR and #12921. I believe this one should be landed first.

**Comment by SeanWykes** [comment]
> As a heads up it looks to me that there are gonna be conflicts between this PR and #12921. I believe this one should be landed first.

Yep, already had a look at #12921 and did what I could to minimize conflicts (such as using r#type to indicate keyType, copied an impl for read_rsa_public_key(), etc). Hopefully should be relatively painless.

**Review by lucacasonato — COMMENTED** [review]
I have a feeling #12921 will land first (probably later today), but I am happy to rebase your PR :-)

File: ext/crypto/00_crypto.js
Comment (SeanWykes) [reply ext/crypto/00_crypto.js:1198]: Done, pls see PR#13027

**Comment by SeanWykes** [comment]
@lucacasonato great work on refactors!

PLMK how you want to proceed with this PR. I think the base refactors have all landed now, so are you still going to rebase this? 

Or should I rework the JWK import on top of main, and submit a new PR? Actually, I am thinking of 3 PRs
- JWK RSA Import
- Refactor exportKey to return the Import enum pkcs8/spki/JWK,raw,
- JWK RSA export. 

Then I can finally rework #13013, hopefully in time for 1.17.

Regards.

**Comment by lucacasonato** [comment]
@SeanWykes Yes, I will rebase and finish this up today. I was just waiting on #13052 to land.

**Comment by lucacasonato** [comment]
And yes, I'll go and pick this PR some more into smaller chunks (first JWK importing). I'll add you as a co-author to all PRs I'll open :-)

**Comment by lucacasonato** [comment]
This PR has now been superseded by #13081 and #13071. Thanks @SeanWykes!

**Comment by SeanWykes** [comment]
@lucacasonato what are your plans for #13013? Can you rebase or should I rework it for the shiny new refactored base?

**Comment by lucacasonato** [comment]
@SeanWykes Up to you: if you want to do it yourself, please split it out into two PRs (one for import, one for export). Otherwise I am also happy to do it 🙂

**Comment by SeanWykes** [comment]
> @SeanWykes Up to you: if you want to do it yourself, please split it out into two PRs (one for import, one for export). Otherwise I am also happy to do it 🙂

I'd like to try, I'm starting to see how thinks tick at deno. I'll split it up as you suggested.

**Pull request closed without merging**
